File: src/experimental/wired-kernel/memory-banks.js
/** Isolated owned digital memory-bank port. Not a board or analog backend. */
import {CircuitFault} from '../digital-circuit.js';
export const MEMORY_BANK_PINS=Object.freeze(['vcc','gnd','oeb','web','select',
    ...Array.from({length:15},(_,i)=>`a${i}`),...Array.from({length:8},(_,i)=>`d${i}`)]);
const SIZE=32768,WORDS=9;
export async function createNativeMemoryBanks({enabled=false,banks,wasmBytes}={}) {
    if(enabled!==true)throw new CircuitFault('EXPERIMENT_DISABLED','enabled:true required');
    if(!Array.isArray(banks)||banks.length<1||banks.length>32)throw new RangeError('native banks 1..32');
    const descriptors=banks.map(({id,kind,contents=new Uint8Array(),readOnly=false},i)=>{
        if(!['62256','28c256'].includes(kind))throw new TypeError('owned 62256/28c256 bank required');
        if(typeof readOnly!=='boolean')throw new TypeError('readOnly');
        if(!(contents instanceof Uint8Array)||contents.length>SIZE)throw new RangeError('bank contents');
        id??=`bank${i}`;if(typeof id!=='string'||!id)throw new TypeError('bank id');
        return {id,kind,readOnly,contents:contents.slice()};
    });
    if(new Set(descriptors.map(b=>b.id)).size!==banks.length)throw new TypeError('duplicate bank id');
    if(!(wasmBytes instanceof Uint8Array))throw new TypeError('owned Wasm bytes required');
    const {instance}=await WebAssembly.instantiate(wasmBytes,{}),e=instance.exports;
    if(e.memory_kernel_version?.()!==1)throw new TypeError('rebuild native memory kernel: ABI version mismatch');
    const count=descriptors.length,start=e.arena_ptr(),capacity=e.arena_capacity(),p={};let end=start;
    const reserve=(name,size)=>{end=Math.ceil(end/4)*4;p[name]=end;end+=size;};
    reserve('memory',count*SIZE);reserve('states',count*WORDS*4);reserve('staged',count*WORDS*4);reserve('protected',count);
    reserve('inputs',count*28);reserve('conflicts',count*28);reserve('drives',count*8);reserve('present',count);reserve('changed',count);reserve('fault',12);
    if(!Number.isSafeInteger(end)||start<0||end-start>capacity||end>e.memory.buffer.byteLength)throw new RangeError('native memory arena capacity');
    const bytes=(name,length)=>new Uint8Array(e.memory.buffer,p[name],length),view=new DataView(e.memory.buffer);
    const pinName=(bank,pin)=>pin===4?(descriptors[bank]?.kind==='62256'?'csb':'ceb'):MEMORY_BANK_PINS[pin]??null;
    const error=(code,bank,pin)=>{
        const codes={1:'FLOATING',2:'UNKNOWN',3:'CONTENTION',4:'MEMORY_POWER',5:'INVALID_DRIVER_LEVEL',6:'INVALID_MEMORY_STATE',
            7:'MEMORY_COUNTER_OVERFLOW',8:'INVALID_BANK_COUNT',9:'INVALID_MEMORY_CONFIG'};
        const id=descriptors[bank]?.id??`bank${bank}`,name=pinName(bank,pin);
        const fault=new CircuitFault(codes[code]??'NATIVE_MEMORY_ERROR',`${id}${name?'.'+name:''}`);fault.bank=id;fault.pin=name;return fault;
    };
    descriptors.forEach((bank,i)=>{
        const memory=new Uint8Array(e.memory.buffer,p.memory+i*SIZE,SIZE);memory.fill(bank.kind==='28c256'?255:0);memory.set(bank.contents);
        view.setUint32(p.states+(i*WORDS+2)*4,0xffffffff,true);bytes('protected',count)[i]=Number(bank.kind==='28c256'&&bank.readOnly);
    });
    const noConflicts=new Uint8Array(count*28);
    const preview=(inputs,conflicts=noConflicts)=>{
        if(!(inputs instanceof Uint8Array)||inputs.length!==count*28||!(conflicts instanceof Uint8Array)||conflicts.length!==inputs.length)
            throw new TypeError('native memory pin dimensions');
        for(let i=0;i<inputs.length;i++)if(inputs[i]>3||conflicts[i]>1||conflicts[i]&&inputs[i]!==2)throw error(5,Math.floor(i/28),i%28);
        bytes('inputs',inputs.length).set(inputs);bytes('conflicts',conflicts.length).set(conflicts);
        const result=e.preview_memory_banks(count,p.memory,p.states,p.staged,p.protected,p.inputs,p.conflicts,p.drives,p.present,p.changed,p.fault);
        if(result)throw error(result,view.getUint32(p.fault+4,true),view.getUint32(p.fault+8,true));
        return descriptors.map((_,i)=>({changed:!!bytes('changed',count)[i],
            drives:bytes('present',count)[i]?new Uint8Array(e.memory.buffer,p.drives+i*8,8).slice():null}));
    };
    const inspect=bank=>{
        if(!Number.isInteger(bank)||bank<0||bank>=count)throw new RangeError('bank index');
        const word=w=>view.getUint32(p.states+(bank*WORDS+w)*4,true),out=word(2);
        return {bytes:new Uint8Array(e.memory.buffer,p.memory+bank*SIZE,SIZE).slice(),cycle:['','idle','read','write'][word(0)],
            addr:word(1),out:out===0xffffffff?-1:out,armed:!!word(3),pending:word(4)?{a:word(5),byte:word(6)}:null,
            writes:word(8)*4294967296+word(7),drives:Uint8Array.from({length:8},(_,i)=>out===0xffffffff?3:(out>>>i)&1)};
    };
    return Object.freeze({capabilities:Object.freeze({experimental:true,digitalMemoryBanks:true,analog:false,cpu:false,board:false,resumableSnapshot:false}),preview,inspect});
}

File: src/experimental/wired-kernel/memory-banks.c
/* Owned ideal-digital SRAM/EEPROM bridge prototype, not an analog solver.
 * State words: cycle, address, output(-1), armed, pending, pending address,
 * pending byte, write-count low/high. Inputs: VCC,GND,OE,WE,SELECT,A0..14,D0..7.
 * The caller supplies disjoint, bounded arrays in its private module arena. */
typedef unsigned int u32;
typedef unsigned char u8;
#define WORDS 9
#define PINS 28
#define SIZE 32768
#define NONE 0xffffffffu
static u32 fail_memory(u32 code,u32 bank,u32 pin,u32 *fault) {
    fault[0]=code;fault[1]=bank;fault[2]=pin;return code;
}
static u32 known_memory(u8 level,u8 conflict,u32 bank,u32 pin,u32 *fault,u8 *value) {
    if(level<2){*value=level;return 0;}
    return fail_memory(level==3?1:conflict?3:2,bank,pin,fault);
}
u32 memory_kernel_version(void){return 1;}
u32 preview_memory_banks(u32 banks,u8 *memory,u32 *states,u32 *staged,const u8 *protected_rom,
                        const u8 *inputs,const u8 *conflicts,u8 *staged_drives,
                        u8 *staged_present,u8 *staged_changed,u32 *fault) {
    if(!banks||banks>32)return fail_memory(8,0,NONE,fault);
    for(u32 b=0;b<banks;b++) {
        if(protected_rom[b]>1)return fail_memory(9,b,NONE,fault);
        for(u32 p=0;p<PINS;p++)if(inputs[b*PINS+p]>3||conflicts[b*PINS+p]>1||(conflicts[b*PINS+p]&&inputs[b*PINS+p]!=2))
            return fail_memory(5,b,p,fault);
        const u32 *s=states+b*WORDS;
        if(s[0]>3||s[1]>=SIZE||(s[2]>255&&s[2]!=NONE)||s[3]>1||s[4]>1||s[5]>=SIZE||s[6]>255||s[8]>0x1fffff)
            return fail_memory(6,b,NONE,fault);
    }
    /* Preflight and preview every bank. No actual state/byte/net-drive commit
     * occurs here; the output buffers are private preview staging as well. */
    for(u32 b=0;b<banks;b++) {
        const u8 *in=inputs+b*PINS,*cf=conflicts+b*PINS;const u32 *s=states+b*WORDS;u32 *n=staged+b*WORDS;
        for(u32 w=0;w<WORDS;w++)n[w]=s[w];
        staged_changed[b]=0;staged_present[b]=1;
        u8 power,oe,we,select=1,bit;u32 error;
        if((error=known_memory(in[0],cf[0],b,0,fault,&power)))return error;
        if(power!=1)return fail_memory(4,b,0,fault);
        if((error=known_memory(in[1],cf[1],b,1,fault,&power)))return error;
        if(power!=0)return fail_memory(4,b,1,fault);
        if((error=known_memory(in[2],cf[2],b,2,fault,&oe)))return error;
        if((error=known_memory(in[3],cf[3],b,3,fault,&we)))return error;
        if((oe!=1||we!=1)&&(error=known_memory(in[4],cf[4],b,4,fault,&select)))return error;
        u32 active=select==0;
        if(!active&&s[0]==1){staged_present[b]=0;continue;}
        u32 address=0,byte=0;
        if(active)for(u32 p=0;p<15;p++) {
            if((error=known_memory(in[5+p],cf[5+p],b,5+p,fault,&bit)))return error;
            address|=(u32)bit<<p;
        }
        if(active&&!we)for(u32 p=0;p<8;p++) {
            if((error=known_memory(in[20+p],cf[20+p],b,20+p,fault,&bit)))return error;
            byte|=(u32)bit<<p;
        }
        u32 cycle=!active?1:!we?3:!oe?2:1;
        if(cycle!=s[0]) {
            n[0]=cycle;n[1]=address;n[2]=NONE;n[4]=n[5]=n[6]=0;
            if(cycle!=3)n[3]=1;
            if(s[0]==3&&s[3]&&s[4]&&!protected_rom[b]) {
                if(s[7]==NONE&&s[8]==0x1fffff)return fail_memory(7,b,NONE,fault);
                n[7]++;if(!n[7])n[8]++;
            }
            staged_changed[b]=1;
        }else if(cycle==1) {
            n[1]=address;n[3]=1;
            if(n[2]!=NONE){n[2]=NONE;staged_changed[b]=1;}
        }else if(cycle==3) {
            n[1]=address;n[4]=1;n[5]=address;n[6]=byte;
        }else {
            byte=memory[b*SIZE+address];
            if(s[1]!=address||s[2]!=byte){n[1]=address;n[2]=byte;staged_changed[b]=1;}
        }
        for(u32 p=0;p<8;p++)staged_drives[b*8+p]=n[2]==NONE?3:(n[2]>>p)&1;
    }
    /* Commit the old pending bytes only after all peer previews succeeded. */
    for(u32 b=0;b<banks;b++) {
        u32 *s=states+b*WORDS,*n=staged+b*WORDS;
        if(s[0]==3&&n[0]!=3&&s[3]&&s[4]&&!protected_rom[b])memory[b*SIZE+s[5]]=(u8)s[6];
        for(u32 w=0;w<WORDS;w++)s[w]=n[w];
    }
    fault[0]=0;return 0;
}
